src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs

```diff
@@ -166,17 +222,58 @@ protected override Status StatusCore
 
         internal Task EndCallAsync()
         {
-            HttpContext.Response.ConsolidateTrailers(this);
+            if (_deadlineLock == null)
+            {
+                EndCallCore();
+                return Task.CompletedTask;
+            }
 
-            if (HasBufferedMessage)
+            var lockTask = _deadlineLock.WaitAsync();
+            if (lockTask.IsCompletedSuccessfully)
             {
-                // Flush any buffered content
-                return HttpContext.Response.BodyWriter.FlushAsync().GetAsTask();
+                try
+                {
+                    EndCallCore();
+                }
+                finally
+                {
+                    _deadlineLock.Release();
+                }
+
+                return Task.CompletedTask;
             }
             else
             {
-                return Task.CompletedTask;
+                return EndCallAsyncCore(lockTask);
             }
+
+            async Task EndCallAsyncCore(Task lockTask)
+            {
+                await lockTask;
+
+                try
+                {
+                    EndCallCore();
+                }
+                finally
+                {
+                    _deadlineLock!.Release();
+                }
+            }
+        }
+
+        private void EndCallCore()
+        {
+            // Don't set trailers if deadline exceeded or request aborted
+            if (CancellationToken.IsCancellationRequested)
+            {
+                return;
+            }
+
+            HttpContext.Response.ConsolidateTrailers(this);
+
+            // Clean up deadline
+            Dispose();
```

Comment:
Does it still make sense for HttpContextServerCallContext to implement `IDisposable` if the user of is expected to call either EndCallAsync or ProcessHandlerError instead?

Reply:
Hmm, probably not. Also now that I think about it instances of this type are passed to the user. Someone could be completely crazy and cast to IDisposable and dispose it.